src/feedcontainer.cpp

```diff
@@ -0,0 +1,226 @@
+	return feed;
+}
+
+void FeedContainer::mark_all_feed_items_read(const unsigned int feed_pos)
+{
+	const auto feed = get_feed(feed_pos);
+	std::lock_guard<std::mutex> lock(feed->item_mutex);
+	std::vector<std::shared_ptr<rss_item>>& items = feed->items();
+	if (items.size() > 0) {
+		bool notify = items[0]->feedurl() != feed->rssurl();
+		LOG(level::DEBUG,
+			"FeedContainer::mark_all_read: notify = %s",
+			notify ? "yes" : "no");
+		for (const auto& item : items) {
+			item->set_unread_nowrite_notify(false, notify);
+		}
+	}
+}
+
+void FeedContainer::add_feed(const std::shared_ptr<rss_feed> feed)
+{
+	std::lock_guard<std::mutex> feedslock(feeds_mutex);
+	feeds.push_back(feed);
+}
+
+void FeedContainer::populate_query_feeds()
+{
+	std::lock_guard<std::mutex> feedslock(feeds_mutex);
+	for (const auto& feed : feeds) {
+		if (feed->rssurl().substr(0, 6) == "query:") {
+			feed->update_items(feeds);
+		}
+	}
+}
+
+unsigned int FeedContainer::get_feed_count_per_tag(const std::string& tag)
+{
+	unsigned int count = 0;
+	std::lock_guard<std::mutex> feedslock(feeds_mutex);
+	for (const auto& feed : feeds) {
+		if (feed->matches_tag(tag)) {
+			count++;
+		}
+	}
+
+	return count;
+}
+
+std::shared_ptr<rss_feed> FeedContainer::get_feed_by_url(
+	const std::string& feedurl)
+{
+	for (const auto& feed : feeds) {
+		if (feedurl == feed->rssurl())
+			return feed;
+	}
+	LOG(level::ERROR, "FeedContainer:get_feed_by_url failed for %s", feedurl);
+	return std::shared_ptr<rss_feed>();
+}
+
+unsigned int FeedContainer::get_pos_of_next_unread(unsigned int pos)
+{
+	std::lock_guard<std::mutex> feedslock(feeds_mutex);
+	for (pos++; pos < feeds.size(); pos++) {
+		if (feeds[pos]->unread_item_count() > 0)
+			break;
+	}
+	return pos;
+}
+
+unsigned int FeedContainer::feeds_size()
+{
+	std::lock_guard<std::mutex> feedslock(feeds_mutex);
+	return feeds.size();
+}
+
+void FeedContainer::reset_feeds_status()
+{
+	std::lock_guard<std::mutex> feedlock(feeds_mutex);
+	for (const auto& feed : feeds) {
+		feed->reset_status();
+	}
+}
+
+void FeedContainer::set_feeds(
+	const std::vector<std::shared_ptr<rss_feed>> new_feeds)
+{
+	std::lock_guard<std::mutex> feedslock(feeds_mutex);
+	feeds = new_feeds;
+}
+
+std::vector<std::shared_ptr<rss_feed>> FeedContainer::get_all_feeds()
+{
+	std::vector<std::shared_ptr<rss_feed>> tmpfeeds;
+	{
+		std::lock_guard<std::mutex> feedslock(feeds_mutex);
+		tmpfeeds = feeds;
+	}
+	return tmpfeeds;
+}
+
+void FeedContainer::clear_feeds_items()
+{
+	std::lock_guard<std::mutex> feedslock(feeds_mutex);
+	for (const auto& feed : feeds) {
+		std::lock_guard<std::mutex> lock(feed->item_mutex);
+		feed->clear_items();
+	}
+}
+
+unsigned int FeedContainer::unread_feed_count()
+{
+	return std::count_if(feeds.begin(),
+		feeds.end(),
+		[](const std::shared_ptr<rss_feed> feed) {
+			return feed->unread_item_count() > 0;
+		});
+}
+
+unsigned int FeedContainer::unread_item_count()
+{
+	unsigned int unread_items = 0;
+	for (const auto& feed : feeds) {
```

Comment:
Just curious: why did you use a standard algorithm in another method but a simple loop here? `std::accumulate` should be suitable for this one.

Reply:
Not sure, couldn't find a suitable method... thanks for being vigilant ;)